### PR TITLE
Fh 4058

### PR DESCRIPTION
**Pull request by philbrookes**
Cross namespace configuration of external services using mcp-api.

Update Mobileservices to note if they are writeable or not.

Update UI to check service is writeable before displaying the create/destroy integration button.

### CODE REVIEW

**Comment by maleck13** [comment]
@philbrookes needs a rebase

File: pkg/openshift/authChecker.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	u.Path = path.Join("/oapi/v1/namespaces/" + namespace + "/localresourceaccessreviews")
+	payload := authCheckJsonPayload{
+		Verb:     "update",
+		Resource: "deploymentconfigs",
+	}
+	bytePayload, err := json.Marshal(payload)
+	if err != nil {
+		return false, errors.Wrap(err, "openshift.ac.Check -> failed to build payload for check authorization")
+	}
+	req, err := http.NewRequest("POST", u.String(), strings.NewReader(string(bytePayload)))
```
Comment (maleck13) [inline pkg/openshift/authChecker.go]: should use ```bytes.NewReader(bytePayload)```

File: pkg/openshift/authChecker.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		return false, errors.Wrap(err, "openshift.ac.Check -> failed to build request to check authorization")
+	}
+	req.Header.Set("authorization", "bearer "+ac.Token)
+	req.Header.Set("Content-Type", "Application/JSON")
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: ac.SkipCertCheck},
+	}
+	client := &http.Client{Transport: tr}
```
Comment (maleck13) [inline pkg/openshift/authChecker.go]: should prob use the clients.HTTPRequesterBuilder

File: pkg/openshift/authChecker.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		}
+
+		return false, errors.New("openshift.ac.Check -> unexpected response code from openshift " + resp.StatusCode)
+	}
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, errors.Wrap(err, "openshift.ac.Check -> failed to read the response body after reading user")
+	}
+	res := &authCheckResponse{}
+	json.Unmarshal(data, res)
```
Comment (maleck13) [inline pkg/openshift/authChecker.go]: need to check errors here

File: pkg/openshift/user.go [hunk 4]
```diff
@@ -39,38 +49,97 @@ func (ua *UserAccess) ReadUserFromToken(host, token string, insecure bool) error
 	client.Timeout = 5 * time.Second
```
Comment (maleck13) [inline pkg/openshift/user.go]: swap out for HTTPRequesterBuilder

**Comment by philbrookes** [comment]
@maleck13 I have tested it, though I'd be happy to go through it with you on a video call to make sure we're both happy with it, if you'd like?

File: pkg/openshift/authChecker.go [hunk 5]
```diff
@@ -100,17 +102,13 @@ func (ac *AuthChecker) Check(resource, namespace string) (bool, error) {
...
 		return false, errors.Wrap(err, "openshift.ac.Check -> failed to build request to check authorization")
 	}
 	req.Header.Set("authorization", "bearer "+ac.Token)
 	req.Header.Set("Content-Type", "Application/JSON")
-	tr := &http.Transport{
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: ac.SkipCertCheck},
-	}
-	client := &http.Client{Transport: tr}
-	client.Timeout = 5 * time.Second
+	client := clients.HttpClientBuilder{}.Insecure(ac.SkipCertCheck).Timeout(5).Build()
```
Comment (maleck13) [inline pkg/openshift/authChecker.go]: heh I meant use it as a dependency so that it can be mocked out

File: pkg/openshift/authChecker.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		return false, errors.Wrap(err, "openshift.ac.Check -> failed to retrieve user details")
+	}
+	u, err := url.Parse(ac.Host)
+	if err != nil {
+		return false, errors.Wrap(err, "openshift.ac.Check -> failed to parse openshift host when attempting to check authorization")
+	}
+	u.Path = path.Join("/oapi/v1/namespaces/" + namespace + "/localresourceaccessreviews")
+	payload := authCheckJsonPayload{
+		Verb:     "update",
+		Resource: "deploymentconfigs",
```
Comment (maleck13) [inline pkg/openshift/authChecker.go]: should this be passed in from the resource var?

File: pkg/openshift/authChecker_test.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+	}
+}
+
+type mockExternalHTTPRequester struct {
+	ResponseBody  io.ReadCloser
+	ResponseError error
+	ResponseCode  int
+}
+
+func (mehr *mockExternalHTTPRequester) Do(req *http.Request) (*http.Response, error) {
```
Comment (maleck13) [inline pkg/openshift/authChecker_test.go]: there is a mock one of these already in the mock package I think

File: pkg/openshift/authChecker_test.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	ac = NewAuthCheckerBuilder(testHost).WithToken(testToken).Build()
+	if reflect.TypeOf(ac).String() != "*openshift.AuthChecker" {
+		t.Fatalf("exected '*openshift.AuthChecker' but got '%s'", reflect.TypeOf(ac).String())
+	}
+	ac = NewAuthCheckerBuilder(testHost).WithToken(testToken).WithUserRepo(testUserRepo).Build()
+	if reflect.TypeOf(ac).String() != "*openshift.AuthChecker" {
+		t.Fatalf("exected '*openshift.AuthChecker' but got '%s'", reflect.TypeOf(ac).String())
+	}
+	ac = NewAuthCheckerBuilder(testHost).IgnoreCerts().Build()
+	if reflect.TypeOf(ac).String() != "*openshift.AuthChecker" {
```
Comment (maleck13) [inline pkg/openshift/authChecker_test.go]: not sure I like the string comparison. I think you can just do ```reflect.TypeOf(ac) != *openshift.AuthChecker```

File: pkg/openshift/user.go [hunk 9]
```diff
@@ -39,38 +49,97 @@ func (ua *UserAccess) ReadUserFromToken(host, token string, insecure bool) error
...
 	data, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return errors.Wrap(err, "failed to read the response body after reading user")
+		return user, errors.Wrap(err, "failed to read the response body after reading user")
 	}
-	ua.Logger.Debug("response from reading user ", string(data))
-	return nil
+
+	userData := &userResponse{}
+	json.Unmarshal(data, userData)
```
Comment (maleck13) [inline pkg/openshift/user.go]: check for error

File: pkg/k8s/mounterBuilder.go [hunk 10]
```diff
@@ -61,6 +62,7 @@ type MountManager struct {
 
 // Mount a secret named mount into the service
 func (mm *MountManager) Mount(service, clientService *mobile.Service) error {
+	fmt.Printf("clientService: %+v\n", clientService)
```
Comment (maleck13) [inline pkg/k8s/mounterBuilder.go]: needs removing

**Pull request merged**